internal/ad/backends/winbind/winbind_test.go

```diff
@@ -0,0 +1,187 @@
+package winbind_test
+
+import (
+	"bytes"
+	"context"
+	"flag"
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"testing"
+
+	"github.com/sirupsen/logrus"
+	"github.com/stretchr/testify/require"
+	"github.com/ubuntu/adsys/internal/ad/backends/winbind"
+	"github.com/ubuntu/adsys/internal/testutils"
+)
+
+func TestWinbind(t *testing.T) {
+	// Build mock libwbclient
+	var mockLibPath string
+	if os.Getenv("GO_WANT_HELPER_PROCESS") != "1" {
+		mockLibPath = buildMockLib(t)
+	}
+
+	// We setup and rerun in a subprocess because we need to preload the mock libwbclient
+	if setupSubprocessForTest(t, mockLibPath) {
+		return
+	}
+
+	tests := map[string]struct {
+		wbclientBehavior string
+		staticADDomain   string
+		staticADServer   string
+
+		wantErr bool
+	}{
+		"Lookup is successful": {},
+
+		// Override cases
+		"Lookup with overridden ad_domain":                  {staticADDomain: "overridden.com"},
+		"Lookup with overridden ad_server":                  {staticADServer: "controller.overridden.com"},
+		"Lookup with overridden ad_server with LDAP prefix": {staticADServer: "ldap://controller.overridden.com"},
+
+		// Error cases
+		"Error looking up domain":      {wbclientBehavior: "domain_not_found", wantErr: true},
+		"Error looking up DC name":     {wbclientBehavior: "error_getting_dc_name"},
+		"Error getting online status":  {wbclientBehavior: "error_getting_online_status"},
+		"Error when domain is offline": {wbclientBehavior: "domain_is_offline"},
+	}
+
+	for name, tc := range tests {
+		tc := tc
+		t.Run(name, func(t *testing.T) {
+			// Set up mock libwbclient behavior
+			t.Setenv("ADSYS_WBCLIENT_BEHAVIOR", tc.wbclientBehavior)
+
+			config := winbind.Config{}
+			if tc.staticADDomain != "" {
+				config.ADDomain = tc.staticADDomain
+			}
+			if tc.staticADServer != "" {
+				config.ADServer = tc.staticADServer
+			}
+
+			winbind, err := winbind.New(context.Background(), config)
```

Comment:
shadowing the package name to a variable could be weird. Maybe we can call it backend?